Rust-lang crate ownership policy (#3119)
rust-lang/rfcs: text/0000-rust-crate-ownership.md

# Rust crate ownership policy

- Feature Name: (fill me in with a unique ident, `my_awesome_feature`)
- Start Date: (fill me in with today's date, 2021-05-04)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

> Have a more intentional policy around crates published by the Rust project, to be applied to existing and future crates published by us.

Reviewer: ```suggestion
Have a more intentional policy around crates published by the Rust project **on crates.io**, to be applied to existing and future crates published by us.
```
There seems to be a lot of confusion in the other comments about the fact that this only applies to crates published on `crates.io`. To reduce confusion, I would suggest that portion of the policy be very clearly and explicitly stated in the summary section. I's stated below in the motivation section, but it's sufficiently important to the policy that putting it in the summary section would be a good idea IMO.

Given that the stamp of an official Rust team carries a degree of weight, it is confusing for community members to have to differentiate between the two, and can lead to incorrect expectations being set. Over a prolonged period of time, this can end up in crates that were never intended to be used widely becoming key dependencies in the ecosystem.

Furthermore, these crates are not necessarily clear on who owns them. Some are owned (in the crates.io sense) by the generic `rust-lang-owner` crates.io account, some are owned by a GitHub team (like `rust-lang/libs`), and yet others are only owned by personal accounts. It seems like we should have some consistency here.

# Reference-Level Explanation

Once accepted, the policy sections of this RFC should be posted on
<https://forge.rust-lang.org/> in a "Rust-lang Crates Policy" section; this RFC will not be the canonical home of the up-to-date crates policy.

## Categories

We propose the following categories of published crates:

>  - **Intentional artifacts**: These are crates which are intentionally released by some team (usually libs), are actively maintained, are intended to be used by external users, and intentionally have an air of officialness. Example: [libc](https://crates.io/crates/libc)

Reviewer: As a reader, the meaning of "intentional" is a needlessly confusing, and "transitively intentional" below even more so. Maybe something like "artifacts for external use"? If you really want this to be the term of art though, it's probably not a huge problem. Just makes the categories a bit more obscure than they need to be.

Author: Yeah "for external use" doesn't flow as well for a term of art, which is why I picked "intentional"

 - **Internal use**: These are crates which are used by some “internal client”, like rustc, crates.io, docs.rs, etc. Their primary purpose is not to be used by external users, though the teams that maintain them (typically the teams of their internal client) may wish for the crate to have wider adoption. The line can be blurry between these and “intentional artifacts” and ultimately depends on the goals of the team. Example: [conduit](https://crates.io/crates/conduit), [measureme](https://crates.io/crates/measureme). There are two subcategories based on whether they are intended to ever show up as a transitive dependency:

>     - **Transitively intentional**: These are dependencies of intentional artifact libraries, and will show up in users' dependency trees, even if they are not intended to be _directly_ used. The Rust Project still needs to handle security issues in these crates _as if_ they are "intentional artifacts".

Reviewer 2: ```suggestion
    - **Transitively intentional**: These are dependencies of intentional artifact libraries, and will show up in users' dependency trees, even if they are not intended to be _directly_ used. The Rust Project still needs to handle security issues in these crates _as if_ they are "intentional artifacts". Example: [compiler_builtins](https://github.com/rust-lang/compiler-builtins).
```

Author: Actually I think that crate is not transitively intentional -- it will never show up in deptrees

Reviewer 2: Hmm, what would you consider a transitively intentional crate then? compiler_builtins is a dependency of core, and will show up if you build core from source with -Zbuild-std.

Author: > If a security issue in one of these crates affects a published binary (or crates.io, etc), that will still need to be handled as a bug in the binary or website.

I should mention the stdlib as well.

A transitively intentional crate would be an internal dependency of `libc`, etc. Basically, it has to show up in your Cargo dependency tree (and as such you need to worry about keeping it up to date).

Author: (I suspect your confusion is stemming from the same place as the other one -- the stdlib is not a _published_ crate, so it is external to this ontology)

 - **Experiment**: This was an experiment by a team, intended to be picked up by users to better inform API design (or whatever), without a long-term commitment to maintainership. Example: [failure](https://crates.io/crates/failure)
 - **Deprecated**: This used to be an “intentional artifact” (or experiment/internal use) but isn’t anymore. Example: [rustc-serialize](https://crates.io/crates/rustc-serialize)
 - **Placeholder**: Not a functional crate, used for holding on to the name of an official tool, etc. Example: [rustup](https://crates.io/crates/rustup)
 - **Expatriated**: This may have been an “intentional artifact”, and still is intended to be used by external users, but is no longer intended to be official. Example: [rand](https://crates.io/crates/rand)

## Policy

> Every crate in the organization must be owned by at least one team on crates.io. Teams should use `rust-lang/foo` teams for this. Non-expatriated crates may not have personal accounts as owners; if a crate needs additional owners that are not part of teams; the team should create a project group. Note that this does not forbid non-team users from having maintainer access to the repository; it simply forbids them from _publishing_.

Reviewer 3: > Note that this does not forbid non-team users from having maintainer access to the repository; it simply forbids them from _publishing_.

When you say "non-team" user here, do you mean "non-team" precisely, or do you mean "non-team/project-group-member" more generally? (I think you mean the latter.)

Author: The latter

> Currently it is not possible for a crate to be owned by _only_ a team; the `rust-lang-owner` account (or a similar account to be decided by the infra team) can be used as a stopgap in such cases, but we should try to phase this account out as much as possible. For crates being auto-published, a `rust-lang/publish-bots` team (or individual bot accounts) can be used to allow bot accounts to publish crates.

Reviewer 4: > For crates being auto-published

Is the intent to have this cover the rustc ap crates as well (e.g. https://crates.io/crates/rustc-ap_rustc_ast)?

Author: Yes!

> Currently it is not possible for a crate to be owned by _only_ a team; the `rust-lang-owner` account (or a similar account to be decided by the infra team) can be used as a stopgap in such cases, but we should try to phase this account out as much as possible. For crates being auto-published, a `rust-lang/publish-bots` team (or individual bot accounts) can be used to allow bot accounts to publish crates.

Reviewer: Maybe add that the reason `rust-lang-owner` is being phased out is to create clarity about who is responsible for the crate?

```suggestion
Currently it is not possible for a crate to be owned by _only_ a team; the `rust-lang-owner` account (or a similar account to be decided by the infra team) can be used as a stopgap in such cases. We should try to phase this account out as much as possible, in order to make sure it is clear who is responsible for each crate. For crates being auto-published, a `rust-lang/publish-bots` team (or individual bot accounts) can be used to allow bot accounts to publish crates.
```

Each crate in the organization, and any future crates in the organization, must decide which to which category they belong in according to the above categorization. If you're not sure what the category should be when registering a crate, or do not wish to make a decision just yet, pick "Experimental".

Each published crate must contain a README. At a minimum, this README must mention the primary owning team. Based on their categories, crates are also required to include the following information in their READMEs and documentation roots:

### Intentional artifact

> “Intentional artifact” crates can choose their commitments but should be clear about what they are in their messaging. If and when a team has a charter, the crate should also be mentioned in the charter as an intentional artifact. Deprecating one of these should not be taken lightly and will require an RFC.

Reviewer 3: When you say "deprecating one of these," do you mean, "deprecating the intentional artifact" itself?

Author: "Deprecating an intentional artifact"

An example of such messaging would be text like:

> This crate is maintained by The Rust \[team\] Team for use by the wider ecosystem. This crate is post-1.0 and follows [semver compatibility](https://doc.rust-lang.org/cargo/reference/semver.html) for its APIs.

Security issues in these crates should be handled with the appropriate weight and careful messaging by the Security Response WG, and should be reported [according to the project's security policy](https://www.rust-lang.org/policies/security).

### Internal use
“Internal use” crates should contain the following text near the top of the readme/documentation:

> > This crate is maintained by \[team\], primarily for use by \[rust project(s)\] and not intended for external use (except as a transitive dependency). This crate may make major changes to its APIs or be deprecated without warning.

Reviewer 2: This is the closest fit for docs.rs and crates.io I think, but the last warning doesn't apply very well to them - they're not libraries, nor will they ever be deprecated. Can the teams have flexibility to change this message a little?

Author: Those are "placeholders" -- if it's not actually a library it's a placeholder; and you should probably not publish the actual code to crates.io. (OTOH `conduit` is an "internal use" library for crates.io)

But yes, I think we can have flexibility.

Reviewer 2: Oh, I think I misunderstood - this is a policy for crates *published to crates.io*, not all crates in the rust-lang org. Never mind then.

Author: Correct.

The "except as a transitive dependency" text should be included if the crate is a dependency of an intentional-artifact library ("transitively intentional").

Security issues in transitively intentional libraries should be handled as if they were intentional artifacts.

### Experiment

“Experiment” crates should mention they are experiments. Experiment crates may be intended to be used in a scoped sort of way; so if they are intended to be used they should be clear about what they are guaranteeing.

An example of such messaging would be text like:

> This crate is maintained by \[team\] as a part of an experiment around \[thingy\]. We encourage people to try to use this crate in their projects and provide feedback through \[method\], but do not guarantee long term maintenance.

or, for experiments that are not intended to be used at all:

> > This crate is maintained by \[team\] and is an internal experiment. We do not guarantee stability or long term maintenance, please use at your own risk.

Reviewer: Minor bikeshed: "please use" reads to me as an encouragement to use the crate, whereas "at your own risk" reads as a discouragement. I think the intention is for it to be a discouragement, so "please use" should maybe be rephrased?

Author: Fixed, it's example text and it doesn't matter that much, and at least in my usage "please use at your own risk" is idiomatic too

Ideally, experimental crates that are published for feedback purposes will have a document to link to that lists out the purpose, rough duration, and processes of the experiment.

### Deprecated
“Deprecated” crates should contain the following text near the top of the readme/documentation:

> This crate is deprecated and not intended to be used.

### Placeholder

“Placeholder” crates should contain the following text in their published readme/documentation:

> This crate is a functionally empty crate that exists to reserve the crate name of \[tool\]. It should not be used. 

> In general it is better to have an empty placeholder crate published instead of reserving the crate via yanking, so that there is a readme that helps

Reviewer: ```suggestion
In general it is better to have an empty placeholder crate published instead of reserving the crate via yanking, so that there is a readme that helps people understand why the crate is unavailable.
```
Even if you don't add this, you should definitely add a period.

### Expatriated

It's unclear if any action should be taken on these beyond removing any semblance of officialness (including rust-lang/foo team owners). We currently have only one such crate (`rand`).

These should by and large not be considered to be "team managed" crates; this category is in this RFC for completeness to be able to talk about expatriation as an end state.

## Transitions and new crates

> Teams should feel free to create new crates in any of these categories; however "Intentional Artifact" crates should ideally be accompanied with an RFC. As we move towards having team charters, this will also involve a charter change. Teams should notify core@rust-lang.org when they've created such crates so that the core team may track these crates and ensure this policy is applied.

Reviewer 5: It'd be great to have some guidelines around intentional crates. To me, they seem to occupy a weird niche—important enough for a team to maintain and publish, but not shipped as part of the toolchain itself. When might this be a sensible solution? Should these crates have a long-term plan to become part of `core` or `std`?

The thrust of this RFC seems to be to impose some structure upon what was a very ad-hoc process of creating semi-official crates in the `rust-lang` org. Given that, the other categories make sense to me, but intentional crates still feel odd. What message are they intended to convey to Rust users? "You can trust and rely on these crates, because they are supported by a Rust team just like `std`"?

If I'm being really honest, crates like `libc` look like missing bits of `core` and `std`. It's pretty hard to write correct FFI without `libc`, and interop via FFI is a huge selling point of Rust. I'd argue that perhaps we should not be encouraging the creation of new crates like this, but acknowledge the ones that currently exist and are important to the Rust ecosystem while determining a path for them to become part of the distributed toolchain.

Author: I'm actually kinda surprised at this question: the distinction, to me, is pretty straightforward and (I thought) well known: things go in std if we're comfortable permastabilizing them, intentional artifact crates outside of std are ones where we expect to use semver to make changes. Rust tends to default to the latter since permastabilizing an API is a pretty large commitment.

> Teams should feel free to create new crates in any of these categories; however "Intentional Artifact" crates should ideally be accompanied with an RFC. As we move towards having team charters, this will also involve a charter change. Teams should notify core@rust-lang.org when they've created such crates so that the core team may track these crates and ensure this policy is applied.

Reviewer 6: > should ideally be accompanied with an RFC.

is this optional? If so can you give example scenarios to show when it would be okay to not create an RFC?

Author: I actually don't think it _should_ be optional, but I was also wary of imposing this on teams? But I do think that an RFC isn't a super high bar if the expectation is that it'll be a small thing.

Also this ties in to the ongoing charter work, _really_ what's being requested here is a change to a charter, which likely will require an FCP, if not an RFC. But the charter stuff is not defined well enough for this RFC to refer to it yet.

Reviewer 6: I'd lean towards just changing it now to say that it requires an RFC and changing it to the charter update process once that is available. I see what you're saying about RFC's being an imposition but how many RFCs like this can we realistically expect between now and when the charter update process is well defined? My money's on 0.

Author: Seems fair. I'm going to leave a hook such that it calls out this potential future update (so updating it isn't a big deal).

From time to time a team's plan for a crate may change: experiments may conclude, crates may need to be deprecated, or the team may decide to release something for wider usage.

In general, teams should notify core@rust-lang.org when such a transition is being made.

Any transition _away_ from "Intentional Artifact" requires an RFC.

> Any transition to "Intentional Artifact" should ideally be accompanied by an RFC.

Reviewer 3: The "transition away" wording is much stronger than "transition to." Is there a particular example you have in mind where a "transition to" _isn't_ accompanied by an RFC?

I note this is very similar to our policy many moons ago. We required an RFC for crates maintained by the libs team before they went to `1.0`.

Author: To me, "transition to" is a charter change, which may not be something which _requires_ an RFC. But I'm open to changing that.

Reviewer 3: Oh I see. Gotya. Just trying to understand better.

Maybe the takeaway for me here is that transitioning both into and out of an "Intentional Artifact" is meant to require a very large speed bump, whether that be an RFC or a change to the charter. That makes sense.

> Expatriation should basically _never_ occur anymore, but it also requires an RFC and core team approval in case it is really necessary.

Reviewer 3: Would it be possible to expand on the strength of wording here? If a team lacks bandwidth to maintain an intentional artifact but some other group of trustworthy people are willing to take it over, then doesn't that require expatriation? Or would y'all expect that the new group become a project team and be brought into the Rust project proper?

(Would it help to perhaps use `rand` as an example? e.g., If this RFC were active back when `rand` moved out of rust-lang, how would/should have it been handled instead?)

Author: > Would it be possible to expand on the strength of wording here? If a team lacks bandwidth to maintain an intentional artifact but some other group of trustworthy people are willing to take it over, then doesn't that require expatriation? Or would y'all expect that the new group become a project team and be brought into the Rust project proper?

I think that either it becomes a team under our governance or it gets forked; expatriation is an option but I think overall the change of guarantees is pretty major.

Reviewer 3: Ah I think I see. It sounds like the fundamental qualification of expatriation is that the same crate name on crates.io is no longer owned by an official Rust group. But it's possible that an "intentional artifact" becomes unmaintained and some other non-official group picks up maintenance, but crucially, with a different crates.io name.

If I have that right, I think it might be worth spelling that out in the RFC text. And I agree, "should basically never happen" sounds about like the right way to handle that case.

Reviewer 7: Moving to a set of trusted maintainers is still a marked shift from "the Rust Project supports this" in my mind - who defines "trusted"? What happens if they become "less trusted"? I think the "silent swap" is the iffy thing for me, and we can just avoid the issue+questions entirely.

For rand:
 - deliberateness around 'intentional artifact' might have prompted a discussion around whether there's real value in 'officialness' and its sustainability long term as a team-supported crate, so it may not have been official in the first place
 - if it had been official, I'd see it going via deprecation and a new crate (like rustc_serialize -> serde)

> Expatriation should basically _never_ occur anymore, but it also requires an RFC and core team approval in case it is really necessary. If a team wishes to stop working on a crate, they should deprecate it and encourage the community to fork it or build their own thing.

Reviewer: What's the reason for this policy? I'm sure there is a good reason, but it would be helpful for readers if it were spelled out in the text.

Reviewer 8: See https://github.com/rust-lang/rfcs/pull/3119#discussion_r625885524.

Reviewer: Thanks! Sorry I missed that. It would definitely be good for it to get added to the text.

Then, working with the teams, we make these changes to their documentation. We also make sure all crates have the appropriate `rust-lang/teamname` github owner, and remove personal accounts from the owners list.

For crates that are in direct use by a lot of the wider community, if we end up categorizing them as anything other than "intentional artifact", there should be an attempt to announce this "change" to the community. While there was no formal commitment made in case of these crates, the vague sense of officialness may have made people believe there was, and we should at least try to rectify this so that people are not continually misled. Whether or not this needs to be done, and how, can be figured out by the individual teams.

A large part of this work can be parallelized; and it does not need to occur all at once.

# Drawbacks
[drawbacks]: #drawbacks

This is a lot of work, but as we move towards a more deliberately structured project, it is probably necessary work.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

> The alternative here is mostly to continue as is. This will become increasingly untenable as we add more and more crates; with the constant danger of internal crates becoming accidental artifacts that the ecosystem depends on.

Reviewer: I'd argue that there's another alternative. You could tell teams to make it clear what level of support they offer for their creates, without standardizing the process or involving the core team. Pros: Less bureaucracy. Cons: No standardization in levels of support across the project, which could make it less clear what category each crate is in. I'm not saying this is a good alternative, but is certainly a better one than the one provided here. In general, I'd suggest that the alternatives presented should be the strongest ones, so that they stimulate discussion during the review process.

Author: Eh, idk. I think "involving core" is not actually the main thrust of this, that's a minor part of this proposal that makes it possible for us to keep track (the proposal still stands without "involving core"). But I'll mention this.

    - **Transitively intentional**: These are dependencies of intentional artifact libraries, and will show up in users' dependency trees, even if they are not intended to be _directly_ used. The Rust Project still needs to handle security issues in these crates _as if_ they are "intentional artifacts".
    - **Not transitively intentional**: These are dependencies of shipped binaries, CI tooling, the stdlib, or are otherwise not expected to show up in users' dependency trees. The Rust Project may need to handle security issues in these crates _internally_, but does not necessarily need to message the wider public about security issues in these crates. If a security issue in one of these crates affects a published binary (or crates.io, etc), that will still need to be handled as a bug in the binary or website.

>  - **Experiment**: This was an experiment by a team, intended to be picked up by users to better inform API design (or whatever), without a long-term commitment to maintainership. Example: [failure](https://crates.io/crates/failure)

Reviewer 9: Is this category for ongoing experiments or finished experiments, or both?

Author: Ongoing, finished ones are deprecated. But it's fine for a finished experiment to still be marked as "experiment", it's not that different from "deprecated" in the guarantees being given.

Security issues in transitively intentional libraries should be handled as if they were intentional artifacts.

### Experiment

“Experiment” crates should mention they are experiments. Experiment crates may be intended to be used in a scoped sort of way; so if they are intended to be used they should be clear about what they are guaranteeing.

An example of such messaging would be text like:

> This crate is maintained by \[team\] as a part of an experiment around \[thingy\]. We encourage people to try to use this crate in their projects and provide feedback through \[method\], but do not guarantee long term maintenance.

or, for experiments that are not intended to be used at all:

> This crate is maintained by \[team\] and is an internal experiment. We do not guarantee stability or long term maintenance, use at your own risk.

Ideally, experimental crates that are published for feedback purposes will have a document to link to that lists out the purpose, rough duration, and processes of the experiment.

### Deprecated
“Deprecated” crates should contain the following text near the top of the readme/documentation:

> > This crate is deprecated and not intended to be used.

Reviewer 9: Does that mean we should publish new versions of crates that have been deprecated to crates.io?

Author: We still can, we are just not _promising_ to keep it up to date

Reviewer 9: Oh, what i meant is if we should publish a new version just to make the new deprecated note appear in the readme on crates.io. Some deprecated crates are marked as such in the repo, but haven't had an update on crates.io since the deprecation.

Author: Oh yeah we should.

### Placeholder

“Placeholder” crates should contain the following text in their published readme/documentation:

> This crate is a functionally empty crate that exists to reserve the crate name of \[tool\]. It should not be used. 

In general it is better to have an empty placeholder crate published instead of reserving the crate via yanking, so that there is a readme that helps people understand why the crate is unavailable.

### Expatriated

> It's unclear if any action should be taken on these beyond removing any semblance of officialness (including rust-lang/foo team owners). We currently have only one such crate (`rand`).

Reviewer 9: > removing any semblance of officialness (including rust-lang/foo team owners)

There are more crates which are not official rust-lang crates that some rust-lang team is co-owner of on crates.io. Does that need to be avoided?

Author: Do you have examples? I think most of these are still "official", e.g. libc is something I consider official.

Reviewer 9: `net2`, `bitflags`, `env_logger`, `simd`. (Depending on the definition of 'official'.)

Author: I would bias towards considering them official but that's a per crate discussion that needs to be had. If it's maintained outside of rust-lang and just happens to have owners it seems ok to consider it unofficial but we should try to be sure.

Author: (by "discussion that needs to be had" I mean that once this rfc lands we will be categorizing the crates with the relevant teams bit by bit, we don't have to pin in down now)
